Merging jenkins plugin test templates into one

jupierce opened the pull request:
When running locally on a developer's system, the previous Jenkins plugin test would conditionally fork and load: 
jenkins-ephemeral-template-test-new-plugin.json
or 
jenkins-ephemeral-template-no-oauth.json

depending on whether the developer had built an image using the OpenShift Jenkins plugin PR-Testing instructions. 

This implementation required that both templates be maintained -- more importantly, failure to do so was undetectable by CI testing. This problem already manifested once when the new-plugin template was not updated to specify the "jenkins" service account. 

Instructing the extended tests to target the local image will also now require an environment variable to be specified before running the tests: USE_SNAPSHOT_JENKINS_IMAGE=1 . This makes explicit the developer's desire to target the local (versus official) jenkins image. A corresponding PR in openshift/jenkins-plugin will be opened to update the README.

@bparees @gabemontero ptal

jupierce commented:
[testextended][extended:core(openshift pipeline plugin)]

gabemontero commented on test/extended/image_ecosystem/plugin.go:
```diff
@@ -414,18 +417,46 @@ var _ = g.Describe("[image_ecosystem][jenkins][Slow] openshift pipeline plugin",
+		useSnapshotImage := os.Getenv(useLocalPluginSnapshotEnvVarName) != "" && snapshotImagePresent
+
+		//TODO disabling oauth until we can update getAdminPassword path to handle oauth (perhaps borrow from oauth integration tests)
+		newAppArgs := []string{exutil.FixturePath("testdata", "jenkins-ephemeral-template-no-oauth.json")}
```
Arguably orthogonal to your change, but if we could somehow copy `examples/jenkins/jenkins-ephemeral-template.json` from the origin source tree to a temporary location (the forcepull.go tests do something similar if I remember correctly), and somehow find this snippet in the template:

```
                  {
                    "name": "OPENSHIFT_ENABLE_OAUTH",
                    "value": "true"
                  },
```
and change the value to `false`, we could in fact use the official example template while still disabling oauth.

Assuming this is sufficiently *simple* (certainly *simpler* I think than handling oauth flows in our extended tests) of course.

Perhaps to be done in a separate PR if we want to get this change in quickly.

bparees replied:
we can make ENABLE_OAUTH a parameter (with a default of true) in the jenkins-ephemeral(and jenkins-persistent) official templates if we wanted to do this, to make it easier.

jupierce replied:
Will do. Definitely the intent of this PR to reduce the templates we need to maintain.

bparees commented on test/extended/image_ecosystem/plugin.go:
```diff
@@ -414,18 +417,46 @@ var _ = g.Describe("[image_ecosystem][jenkins][Slow] openshift pipeline plugin",
+			ginkgolog("")
+
+			// Create an imagestream based on the Jenkins' plugin PR-Testing image (https://github.com/openshift/jenkins-plugin/blob/master/PR-Testing/README).
+			err = oc.Run("new-build").Args("-D", "FROM openshift/jenkins-plugin-snapshot-test", "--to", "jenkins-plugin-snapshot-test").Execute()
```
make the FROM image name a constant and use the same constant on line 418 (the "tag" array).

i don't think there should be any issue with having ":latest" on the image name in both cases.

gabemontero commented:
On Tue, Nov 1, 2016 at 3:04 PM, Ben Parees <notifications@github.com> wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In test/extended/image_ecosystem/plugin.go
> <https://github.com/openshift/origin/pull/11711>:
>
> > @@ -414,18 +417,46 @@ var _ = g.Describe("[image_ecosystem][jenkins][Slow] openshift pipeline plugin",
>  		g.By("kick off the build for the jenkins ephermeral and application templates")
>  		tag := []string{"openshift/jenkins-plugin-snapshot-test:latest"}
>  		hexIDs, err := exutil.DumpAndReturnTagging(tag)
> -		var jenkinsEphemeralPath string
> -		var testingSnapshot bool
> -		if len(hexIDs) > 0 && err == nil {
> -			// found an openshift pipeline plugin test image, must be testing a proposed change to the plugin
> -			jenkinsEphemeralPath = exutil.FixturePath("testdata", "jenkins-ephemeral-template-test-new-plugin.json")
> -			testingSnapshot = true
> +
> +		// If the user has expressed an interest in local plugin testing by setting the
> +		// SNAPSHOT_JENKINS_IMAGE environment variable, try to use the local image. Inform them
> +		// either about which image is being used in case their test fails.
> +		snapshotImagePresent := len(hexIDs) > 0 && err == nil
> +		useSnapshotImage := os.Getenv(useLocalPluginSnapshotEnvVarName) != "" && snapshotImagePresent
> +
> +		//TODO disabling oauth until we can update getAdminPassword path to handle oauth (perhaps borrow from oauth integration tests)
> +		newAppArgs := []string{exutil.FixturePath("testdata", "jenkins-ephemeral-template-no-oauth.json")}
>
> we can make ENABLE_OAUTH a parameter (with a default of true) in the
> jenkins-ephemeral(and jenkins-persistent) official templates if we wanted
> to do this, to make it easier.
>
True, but I thought when discussed in the past we liked the env var
apporach over template parameter for other reasons.  Would this scenario be
sufficient to change are thoughts here?

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11711>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadDcC7NY0pZK0rwFfPZaKFrK__rLlks5q540ygaJpZM4KmcB0>
> .
>

bparees commented:
> True, but I thought when discussed in the past we liked the env var apporach over template parameter for other reasons. Would this scenario be sufficient to change are thoughts here?

@gabemontero it's still an env var to the container, it's just that now the env var value would be controllable by a template parameter instead of being hardcoded in the template.

gabemontero commented:
On Tue, Nov 1, 2016 at 3:26 PM, Ben Parees <notifications@github.com> wrote:

> True, but I thought when discussed in the past we liked the env var
> apporach over template parameter for other reasons. Would this scenario be
> sufficient to change are thoughts here?
>
> @gabemontero <https://github.com/gabemontero> it's still an env var to
> the container, it's just that now the env var value would be controllable
> by a template parameter instead of being hardcoded in the template.
>
@bparees  I get that, but I thought I remembered during our discussions a
perceived advantage to having the env var hardcoded in this particular
case.  Essentially we didn't want to be
"outgoing" in allowing users to override oauth.  But perhaps I'm
misremembering.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11711#issuecomment-257667909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadH6XaDSBZyJCcdJi9eWufuZhunwmks5q55J-gaJpZM4KmcB0>
> .
>

bparees commented:
> Essentially we didn't want to be "outgoing" in allowing users to override oauth. But perhaps I'm misremembering.

i think we may have felt it didn't need to be a parameter, but i can't think of a good reason not to allow it to be a parameter, aside from potential user confusion about whether they should be changing it.

if they set it to false, they are going to get admin/password behavior, which i guess is undesirable, but we can note that in the parameter description.

gabemontero commented:
On Tue, Nov 1, 2016 at 3:53 PM, Ben Parees <notifications@github.com> wrote:

> Essentially we didn't want to be "outgoing" in allowing users to override
> oauth. But perhaps I'm misremembering.
>
> i think we may have felt it didn't need to be a parameter, but i can't
> think of a good reason not to allow it to be a parameter, aside from
> potential user confusion about whether they should be changing it.
>
> if they set it to false, they are going to get admin/password behavior,
> which i guess is undesirable, but we can note that in the parameter
> description.
>
OK then.  I'll jump on board then.  And we want @jupierce to make that
change as part of this PR ?  If so, don't forget
jenkins-persistent-template.json as well.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11711#issuecomment-257675703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadM_SRZETnlBBovxGjNSV_BqJDCWpks5q55i0gaJpZM4KmcB0>
> .
>

bparees commented:
> OK then.  I'll jump on board then.  And we want @jupierce to make that
change as part of this PR ?

might as well, if we're consolidating templates let's go all the way.

jupierce commented:
- String constants used.
- Removed no-oauth template. 
- Parameterized official jenkins ephemeral/persistent templates with ENABLE_OAUTH.
- Changed test to use official ephemeral template. 
@bparees @gabemontero ptal

gabemontero commented:
IGTM

bparees commented on test/extended/image_ecosystem/plugin.go:
```diff
@@ -412,20 +418,49 @@ var _ = g.Describe("[image_ecosystem][jenkins][Slow] openshift pipeline plugin",
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			// Supplant the normal imagestream with the local imagestream using template parameters
+			newAppArgs = append(newAppArgs, "-p", fmt.Sprintf("NAMESPACE=%s,JENKINS_IMAGE_STREAM_TAG=%s:latest", oc.Namespace(), snapshotImageStream))
```
needs ENABLE_OAUTH=false
also split these parameters into multiple -p args, because we're disabling the comma separated key/value pairs behavior: https://github.com/openshift/origin/pull/11539

bparees replied:
sorry, i see you're handling OAUTH globally above.  still need to use separate -p args instead of comma separated though.

jupierce replied:
ENABLE_OAUTH is set to false for all codepaths before this point (https://github.com/openshift/origin/pull/11711/files#diff-e4304911bb619db276427757be12a8c5R431) .
Comma change in next push.

bparees commented:
still needs hack/update-generated-bootstrap.sh, otherwise lgtm.

jupierce commented:
- Commas no longer user for multiple template parameters.
- Generated boostrap 
- If dev defines USE_SNAPSHOT_JENKINS_IMAGE but does not have local jenkins snapshot, the test will fail.

bparees commented:
lgtm

jupierce commented:
Flake #11735
[test]

jupierce commented:
[testextended][extended:core(openshift pipeline plugin)]

bparees commented:
[merge]

bparees commented:
[merge]

Ben Parees | OpenShift

On Nov 3, 2016 12:53 AM, "OpenShift Bot" <notifications@github.com> wrote:

> continuous-integration/openshift-jenkins/merge FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/11041/) (Base
> Commit: 75ed7e1
> <https://github.com/openshift/origin/commit/75ed7e1c18855ea7fb0997043851057ba65faf76>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/11711#issuecomment-258064367>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEvl3tt3cconXFeyVn9d7M24rjaaB1h9ks5q6WjhgaJpZM4KmcB0>
> .
>

bparees commented:
flake https://github.com/openshift/origin/issues/11650
[merge]
